File: src/main/java/net/onelitefeather/vulpes/generator/generation/java/ItemGenerator.java
package net.onelitefeather.vulpes.generator.generation.java;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import io.micronaut.context.annotation.Prototype;
import jakarta.inject.Inject;
import net.kyori.adventure.key.Key;
import net.kyori.adventure.text.Component;
import net.minestom.server.MinecraftServer;
import net.minestom.server.item.ItemStack;
import net.minestom.server.item.Material;
import net.minestom.server.item.component.EnchantmentList;
import net.minestom.server.item.enchant.Enchantment;
import net.minestom.server.registry.RegistryKey;
import net.onelitefeather.vulpes.api.model.ItemEntity;
import net.onelitefeather.vulpes.api.model.item.ItemEnchantmentEntity;
import net.onelitefeather.vulpes.api.repository.ItemRepository;
import net.onelitefeather.vulpes.generator.generation.AbstractCodeGenerator;
import net.onelitefeather.vulpes.generator.util.MaterialParser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.lang.model.element.Modifier;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static net.onelitefeather.vulpes.generator.util.Constants.INDENT_DEFAULT;

@Prototype
public final class ItemGenerator extends AbstractCodeGenerator<ItemEntity> implements JavaStructure {

    private final ItemRepository itemRepository;

    @Inject
    public ItemGenerator(@NotNull ItemRepository itemRepository) {
        super("ItemRegistry", BASE_PACKAGE + ".item");
        this.itemRepository = itemRepository;
    }

    @Override
    protected List<ItemEntity> getModels() {
        return this.itemRepository.findAll();
    }

    @Override
    public void generate(@NotNull Path javaPath) {
        List<ItemEntity> models = getModels();

        if (models.isEmpty()) return;

        // this.classSpec.addJavadoc(defaultDocumentation)
        addClassModifiers(this.classBuilder);
        addJetbrainsAnnotation(this.classBuilder);
        addPrivateDefaultConstructor(this.classBuilder);
        addDefaultSuppressAnnotation(this.classBuilder);
        Map<String, FieldSpec> itemFields = new HashMap<>();

        ClassName itemStackClass = ClassName.get(ItemStack.class);
        ClassName materialClass = ClassName.get(Material.class);

        models.stream().filter(itemModel -> {
                    String variableName = itemModel.getVariableName();
                    return variableName != null && !variableName.isEmpty();
                })
                .forEach(model -> {
                    if (itemFields.containsKey(model.getVariableName())) return;
                    Material material = MaterialParser.fromKey(model.getMaterial());

                    var initBlock = CodeBlock.builder();

                    initBlock.add(
                            "\\$T.builder(\\$T.\\$L)",
                            itemStackClass,
                            materialClass,
                            material
                    );
                    if (model.getAmount() != 0) {
                        initBlock.addStatement(".amount(\\$L)", model.getAmount());
                    }

                    if (model.getDisplayName() != null) {
                        Component parsedName = Component.text(model.getDisplayName());
                        initBlock.addStatement(".customName(\\$L)", parsedName);
                    }

                    if (!model.getEnchantments().isEmpty()) {
                        Map<RegistryKey<@NotNull Enchantment>, Integer> enchantmentData = new HashMap<>();

                        for (ItemEnchantmentEntity enchantment : model.getEnchantments()) {
                            Key key = Key.key(enchantment.getName());
                            RegistryKey<Enchantment> minecraftEnchantment = MinecraftServer.getEnchantmentRegistry().getKey(key);
                            enchantmentData.put(minecraftEnchantment, ((int) enchantment.getLevel()));
                        }
                        EnchantmentList enchantmentList = new EnchantmentList(enchantmentData);
                        initBlock.addStatement(".set(DataComponents.ENCHANTMENTS, \\$L)", enchantmentList);
                    }
                    initBlock.add(".build()");
                    itemFields.put(model.getVariableName(), FieldSpec.builder(itemStackClass, model.getVariableName())
                            .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
                            .initializer(initBlock.build())
                            .build());
                });

        filesToGenerate.clear();
        this.classBuilder.addFields(itemFields.values());
        filesToGenerate.add(
                JavaFile.builder(packageName, this.classBuilder.build())
                        .indent(INDENT_DEFAULT)
                        .build()
        );
        writeFiles(javaPath);
    }

    @Override
    public @NotNull String getName() {
        return "ItemGenerator";
    }

    /**
     * Generates the item fields for the given models.
     *
     * @param itemModel the models to generate the fields for
     * @return the generated fields
     */
    private @NotNull String getDisplayName(@NotNull ItemEntity itemModel) {
        var displayName = itemModel.getDisplayName() == null ? "" : itemModel.getDisplayName();
        return displayName.isEmpty() ? "No display name provided for " + itemModel.getUiName() : displayName;
    }

    /**
     * Builds the lore for the given item model.
     *
     * @param lore the lore to build
     * @return the generated lore
     */
    private @Nullable CodeBlock buildLore(@NotNull List<String> lore) {
        if (lore.isEmpty()) return null;
        var loreBlock = CodeBlock.builder();
        ClassName componentClass = ClassName.get(Component.class);
        int lastEntry = lore.size() - 1;

        loreBlock.add(".lore(\\$T.of(");
        for (int i = 0; i < lore.size(); i++) {
            loreBlock.add("\\$T.text(\\$S)", componentClass, lore.get(i));
            if (i != lastEntry) {
                loreBlock.add(", ");
            }
        }
        loreBlock.add("))");
        return loreBlock.build();
    }

    private @NotNull CodeBlock buildEnchantments(@NotNull Map<String, Short> enchantments) {
        var enchantmentBlock = CodeBlock.builder();
        for (Map.Entry<String, Short> entry : enchantments.entrySet()) {
            enchantmentBlock.add(".enchantment(");
            enchantmentBlock.add("\\$T.fromNamespaceId(\\$S)", ClassName.get(Enchantment.class), entry.getKey());
            enchantmentBlock.add(", (short) \\$L)", entry.getValue());
        }
        return enchantmentBlock.build();
    }
}

File: src/main/kotlin/net/theevilreaper/vulpes/generator/generation/java/ItemGenerator.kt
package net.theevilreaper.vulpes.generator.generation.java

import com.squareup.javapoet.CodeBlock
import com.squareup.javapoet.FieldSpec
import com.squareup.javapoet.JavaFile
import jakarta.inject.Inject
import jakarta.inject.Singleton
import net.kyori.adventure.text.Component
import net.minestom.server.item.Enchantment
import net.minestom.server.item.ItemHideFlag
import net.minestom.server.item.ItemStack
import net.minestom.server.item.Material
import net.theevilreaper.vulpes.api.model.ItemModel
import net.theevilreaper.vulpes.api.repository.ItemRepository
import net.theevilreaper.vulpes.api.util.*
import net.theevilreaper.vulpes.generator.generation.BaseGenerator
import net.theevilreaper.vulpes.generator.util.BASE_PACKAGE
import net.theevilreaper.vulpes.generator.util.JavaGenerationHelper
import net.theevilreaper.vulpes.generator.util.INDENT_DEFAULT
import net.theevilreaper.vulpes.generator.util.META_DATA_VARIABLE
import java.nio.file.Path

/**
 * @author theEvilReaper
 * @version 1.0.0
 * @since
 **/
@Singleton
class ItemGenerator @Inject constructor(
    val itemRepository: ItemRepository
) : JavaGenerationHelper, BaseGenerator<ItemModel>(
    className = "ItemRegistry",
    packageName = "$BASE_PACKAGE.item",
) {
    override fun generate(javaPath: Path) {
        val models = getModels()

        if (models.isEmpty()) return

        this.classSpec.addJavadoc(defaultDocumentation)
        addClassModifiers(this.classSpec)
        addJetbrainsAnnotation(this.classSpec)
        addPrivateDefaultConstructor(this.classSpec)
        addSuppressAnnotation(this.classSpec)
        val itemFields: MutableMap<String, FieldSpec> = mutableMapOf()

        models.filter { it.name.orEmpty().trim().isNotEmpty() }.forEach { model ->
            //Models can be ignored when the namespace aka the material is null or include not the valid namespace
            if (itemFields.isNotEmpty() && itemFields.containsKey(model.name!!)) return@forEach

            val material = if (model.hasMaterial()) {
                Material.fromNamespaceId(model.material!!) ?: Material.STONE
            } else {
                Material.STONE
            }
            val init = CodeBlock.builder()

            init.add(
                "\$T.builder(\$T.\$L)",
                ItemStack::class.java,
                Material::class.java,
                material.namespace().path().uppercase(),
            )

            if (model.amount != null) {
                init.add(".amount(\$L)", model.getAmount())
            }

            if (model.hasMetaData()) {
                val metadata = CodeBlock.builder()

                metadata.add(".meta($META_DATA_VARIABLE -> ")
                metadata.indent()
                metadata.add("$META_DATA_VARIABLE.")

                if (model.customModelId != null) {
                    metadata.add("customModelData(\$L)", model.customModelId)
                }

                if (model.hasDisplayName()) {
                    metadata.add(".displayName(\$T.text(\$S))", Component::class.java, getDisplayName(model))
                }

                if (model.hasFlags()) {
                    metadata.add(buildItemFlags(model.flags!!))
                }

                if (model.hasLoreLines()) {
                    metadata.add(buildLore(model.lore!!))
                }

                if (model.hasEnchantments()) {
                    metadata.add(buildEnchantments(model.enchantments!!))
                }

                metadata.unindent()
                metadata.add(")")
                init.add(metadata.build())
            }
            init.add(".build()")

            itemFields[model.name!!] = FieldSpec.builder(ItemStack::class.java, model.name!!.uppercase())
                .addModifiers(*defaultModifiers)
                .initializer(init.build())
                .build()

        }
        filesToGenerate.clear()
        this.classSpec.addFields(itemFields.values)
        filesToGenerate.add(
            JavaFile.builder(packageName, this.classSpec.build()).indent(INDENT_DEFAULT).build()
        )
        writeFiles(javaPath)
    }

    override fun getName(): String = "ItemGenerator"

    override fun getModels(): List<ItemModel> = itemRepository.findAll()

    private fun getDisplayName(model: ItemModel): String {
        val displayName = model.displayName.orEmpty()
        return when(displayName.isNotEmpty()) {
            true -> displayName
            false -> "No display name provided for ${model.name}"
        }
    }

    private fun buildLore(lore: List<String>): CodeBlock {
        val block = CodeBlock.builder()
        val lastEntry = lore.size - 1
        block.add(".lore(\$T.of(")
        lore.forEachIndexed { index, loreEntry ->
            block.add("\$T.text(\$S)", Component::class.java, loreEntry)
            if (index < lastEntry) {
                block.add(",")
            }
        }
        block.add("))")
        return block.build()
    }

    private fun buildEnchantments(enchantments: Map<String, Short>): CodeBlock {
        val block = CodeBlock.builder()
        enchantments.entries.forEach { entry ->
            block.add(".enchantment(")
            block.add("\$T.fromNamespaceId(\$S)", Enchantment::class.java, entry.key)
            block.add(", (short) \$L", entry.value)
        }

        return block.build()
    }

    private fun buildItemFlags(itemFlags: List<String>): CodeBlock {
        val codeBlock = CodeBlock.builder()
        codeBlock.add(".hideFlag(")
        val lastEntry = itemFlags.size - 1
        itemFlags.forEachIndexed { index, flag ->
            codeBlock.add("\$T.\$L", ItemHideFlag::class.java, ItemHideFlag.valueOf(flag).name)
            if (index < lastEntry) {
                codeBlock.add(", ")
            }
        }
        codeBlock.add(")")
        return codeBlock.build()
    }
}
